### PR TITLE
Minor Bug Fixes

### PR DESCRIPTION
**Pull request by WilliamIX**
This pull request fixes a some issues on the issues page including:

https://github.com/BuildACell/BioCRNPyler/issues/115 (fixing leak in Repressible and Activatable Promoter classes)

https://github.com/BuildACell/BioCRNPyler/issues/92 (fixing filter_dict issues in GlobalMechanisms). New tests and documentation were added here as well.

And making sure build_crns_directly.xml is ignored.

### CODE REVIEW

File: biocrnpyler/chemical_reaction_network.py [hunk 1]
```diff
@@ -232,6 +236,18 @@ def replace_species(self, species: Species, new_species: Species):
         
         return ComplexSpecies(species = new_species_list, name = new_name, material_type = self.material_type, attributes = self.attributes)
 
+    #Returns all species in the ComplexSpecies. If recursive = True, returns species inside internal ComplexSpecies recursively as well.
+    def get_species(self, recursive = False):
```
Comment (zoltuz) [inline biocrnpyler/chemical_reaction_network.py:240]: Is there a strong case for recursive=True? I plan to change this to a more pythonic code in #103, namely:
```python
class ComplexSpecies:
     @property 
     def species(self):
         return self.species
```    
Then, you can use it like:
```python
cp = ComplexSpecies([sp1,sp2])
print(cp.species)
```
You don't need to change the code, that's not my point, just I want to know why recursive=True is needed right now.

File: biocrnpyler/dna_assembly_promoter.py [hunk 2]
```diff
@@ -108,11 +108,13 @@ def update_reactions(self):
 
 #A class for a promoter which can be activated by a single species, modelled as a positive hill function
 class ActivatablePromotor(Promoter):
-    def __init__(self, name, activator, transcript = None, **keywords):
+    def __init__(self, name, activator, leak = False, transcript = None, **keywords):
```
Comment (zoltuz) [inline biocrnpyler/dna_assembly_promoter.py:111]: This is fine for now, but this change breaks all the code which use the old format with positional arguments:
```python
ap = ActivatablePromotor(name,activator,transcript)
```
I guess this class is not used extensively right now, so it's fine for now

File: biocrnpyler/dna_assembly_promoter.py [hunk 3]
```diff
@@ -125,25 +127,31 @@ def update_species(self, **keywords):
...
+
         reactions += mech_tx.update_reactions(dna = self.assembly.dna, transcript = self.transcript, regulator = self.activator, 
-                                             component = self, part_id = self.name+"_"+self.activator.name, **keywords)
+                                             component = self, part_id = self.name+"_"+self.activator.name, leak = self.leak, **keywords)
+
+        
         return reactions
 
 #A class for a promoter which can be repressed by a single species, modelled as a negative hill function
 class RepressablePromotor(Promoter):
```
Comment (zoltuz) [inline biocrnpyler/dna_assembly_promoter.py:147]: RepressiblePromoter

File: biocrnpyler/dna_assembly_promoter.py [hunk 4]
```diff
@@ -108,11 +108,13 @@ def update_reactions(self):
 
 #A class for a promoter which can be activated by a single species, modelled as a positive hill function
 class ActivatablePromotor(Promoter):
```
[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/dna_assembly_promoter.py:110]: whoops that would be me. Will change.

File: biocrnpyler/chemical_reaction_network.py (hunk 1 above)
Comment (WilliamIX) [reply biocrnpyler/chemical_reaction_network.py:240]: So I think this is a little different than just accessing ComplexSpecies.species.

In the non-recrusive case, only the ComplexSpecies is returned. In the recursive case, all the Species in the ComplexSpecies are returned and furthermore, if the Species inside the ComplexSpecies is also a ComplexSpecies, the species inside that ComplexSpecies are returned and so on recursively.

I think this actually makes 3 cases, which maybe we can have different functions for (or the same function with different keywords):

1. Get the list of species inside a ComplexSpecies, with no kind of recursion.
2. Get the list of species inside a ComplexSpecies with recursion (ie flatten all nested ComplexSpecies).
3. A base-case for 2 so when a Species is hit the recursive function stops.

In terms of this code, it is important to have both recursive and non-recursive calls in order to filter ComplexSpecies correctly in different cases when using GlobalMechanisms. I can also imagine other cases when we might want to flatten all nested ComplexSpecies.

File: biocrnpyler/dna_assembly_promoter.py (hunk 2 above)
Comment (WilliamIX) [reply biocrnpyler/dna_assembly_promoter.py:111]: I will fix this too - good catch thanks!

File: biocrnpyler/dna_assembly_promoter.py [hunk 5]
```diff
@@ -125,25 +127,31 @@ def update_species(self, **keywords):
...
-                                             component = self, part_id = self.name+"_"+self.activator.name, **keywords)
+                                             component = self, part_id = self.name+"_"+self.activator.name, leak = self.leak, **keywords)
+
+        
         return reactions
 
 #A class for a promoter which can be repressed by a single species, modelled as a negative hill function
-class RepressablePromotor(Promoter):
-    def __init__(self, name, repressor, transcript = None, **keywords):
+class RepressablePromoter(Promoter):
```
Comment (zoltuz) [inline biocrnpyler/dna_assembly_promoter.py:147]: Sorry for being a grammar nazi, but I think it is called a RepressiblePromoter
e.g. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3599224/

Comment (WilliamIX) [reply biocrnpyler/dna_assembly_promoter.py]: Sorry for being a bad speller. Typo-corrected.

Comment (zoltuz) [reply biocrnpyler/dna_assembly_promoter.py]: no worries, just this PR it the perfect place to weed out this typo.  Thanks for taking care of it!

**Pull request merged**